themes/src/main/resources/theme/keycloak.v2/account/src/app/content/signingin-page/SigningInPage.tsx

```diff
@@ -249,7 +259,32 @@ class SigningInPage extends React.Component<SigningInPageProps, SigningInPageSta
 
     private credentialRowCells(credential: UserCredential, type: string): React.ReactNode[] {
         const credRowCells: React.ReactNode[] = [];
-        credRowCells.push(<DataListCell id={`${SigningInPage.credElementId(type, credential.id, 'label')}`} key={'userLabel-' + credential.id}>{credential.userLabel}</DataListCell>);
+        const credData: CredData = JSON.parse(credential.credentialData!);
+        credRowCells.push(
+            <DataListCell id={`${SigningInPage.credElementId(type, credential.id, 'label')}`} key={'userLabel-' + credential.id}>
+                {credential.userLabel}
+                {credData.remainingCodes && credData.totalCodes &&
+                    <div>{credData.totalCodes - credData.remainingCodes}/{credData.totalCodes} recovery codes used</div>
+                }
+                {credData.remainingCodes && credData.remainingCodes < 4 &&
+                    <>
+                        <br />
+                        <div className="pf-c-alert pf-m-warning pf-m-inline" aria-label="Success alert">
+                            <div className="pf-c-alert__icon">
+                                <i className="pficon-warning-triangle-o" aria-hidden="true"></i>
+                            </div>
+                            <h4 className="pf-c-alert__title">
```

Comment:
It will be ideal if we can avoid the account-console code specific for "recovery-code" credential type.

The account console is supposed to be used with custom credential type implementations provided by administrator. Having the code specific for each credential does not scale IMO. Will be good if we can handle all of this with credential metadata.

Reply:
I can move the userLabel generation inside the code. I am not sure how to call the alert in agnostic way.

Reply:
@stelewis-redhat Yes, the second thing you mentioned. I am thinking about CredentialMetadata as separate object, which will contain "CredentialModel" as one of the fields. See my comment above.

The main motivation is that `SigningInPage.tsx` doesn't need to contain any if/else blocks or other hacks specific to `Recovery-code` credential. As it will be ideal that customers, who potentially implement their own credentials/authenticators don't need to change UI classes, but SPI is good enough so they can add their own info/warning labels shown alongside their credential just by implementing this method on `CredentialProvider` interface.

Reply:
OK. I can switch that around. I have a working solution for the first, but it shouldn't be too hard to swap that structure.

Reply:
@mposolda Is it possible to revisit making CredentialMetadata contain CredentialModel, after we do the account console redesign? We're a bit worried about the merge/rebasing issues with these two initiatives going on at once.

Here's what I currently have working, for reference: https://github.com/rh-iatanaso/keycloak/pull/14 not final, work in progress.

Reply:
I vote to avoid doing changes in `CredentialModel` and `CredentialRepresentation` classes and rather do it the other way around.
The problem with changing `CredentialModel` is, that the class is part of public SPI and part of the model. This mean that if we change anything in it, it will become part of "public SPI" and it is a pain to remove/change it later. So ideal would be if the initial commit already is done in a way that `CredentialModel` does not need to be changed.

Reply:
OK. Thank you @mposolda